### PR TITLE
344 staking contract v2

### LINKED ISSUE #344: Staking Contract
For the upcoming oStake Module we will need to implement a staking Contract.
The contract holds balances and allows for deposit/withdrawal (also of the depositFor() kind).

Possibly the easiest implementation for the staking use case would be to structure it as an [ERC20Snapshot](https://docs.openzeppelin.com/contracts/3.x/api/token/erc20#ERC20Snapshot) that gives out non-transferable receipt tokens (can be thought as transferable for future). This would allow us to query user balance based on a timestamp, which we can then use for reward distribution in other places.

### PR DESCRIPTION
**Pull request by FHieser**
Solves #344

### CODE REVIEW

**Comment by 0xNuggan** [comment]
Everything looks good, but I think it would be good if you added descriptions to the storage variables

File: src/modules/logicModule/StakingManager.sol [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+    modifier validDuration(uint duration) {
+        if (duration == 0) {
+            revert Module__StakingManager__InvalidDuration();
+        }
+        _;
+    }
+
+    //--------------------------------------------------------------------------
+    // Storage
```
Comment (0xNuggan) [inline src/modules/logicModule/StakingManager.sol:44]: Maybe add some descriptions here

**Review by 0xNuggan — CHANGES_REQUESTED** [review]
Everything looks good, but I think it would be good if you added descriptions to the storage variables

File: src/modules/logicModule/StakingManager.sol (hunk 1 above)
Comment (FHieser) [reply src/modules/logicModule/StakingManager.sol:44]: Done

**Review by 0xNuggan — APPROVED** [review]

**Pull request merged**